agent/agents/supervisor/supervisor.go

```diff
@@ -213,6 +213,32 @@ func (s *Supervisor) SetState(state *agentpb.SetStateRequest) {
 	s.setBuiltinAgents(state.BuiltinAgents)
 }
 
+// RestartAgents restarts all existing agents.
+func (s *Supervisor) RestartAgents() {
+	s.rw.Lock()
+	defer s.rw.Unlock()
+
+	for id, agent := range s.agentProcesses {
+		agent.cancel()
+		<-agent.done
+
+		if err := s.startProcess(id, agent.requestedState, agent.listenPort); err != nil {
+			s.l.Errorf("Failed to restart Agent: %s.", err)
+			// TODO report that error to server
+		}
+	}
+
+	for id, agent := range s.builtinAgents {
+		agent.cancel()
+		<-agent.done
+
+		if err := s.startBuiltin(id, agent.requestedState); err != nil {
+			s.l.Errorf("Failed to restart Agent: %s.", err)
+			// TODO report that error to server
+		}
+	}
+}
```

Comment:
Is it possible to use existing functions `setAgentProcesses` and `setBuiltinAgents`? If those functions are too big it's better to split them into smaller ones. Otherwise we're duplicating the code and the TODOs.

Reply:
These functions restart agents only when state was changed, while we need to restart agents without any change.